### PR TITLE
[mlir][gpu] Skip address space checks for memrefs between launchFuncOp and kernel func

### PR DESCRIPTION
**Pull request by kurapov-peter**
Hi! I've been trying to use the new signature lowering (https://github.com/llvm/llvm-project/pull/101664 and https://github.com/llvm/llvm-project/pull/102621) with gpu kernel outlining, gpu binary generation, and OpenCL runtime. `gpu-to-llvm-spv` pass can handle memrefs with address spaces attributes, however, when lowering an arbitrary function (say, a function that accepts a tensor and then runs bufferization) and then using its argument as a kernel argument the address space is lacking.

Consider the following example function that accepts some device-host shared memory:
```mlir
  func.func @foo(%mem : memref<5xf32>) {
    gpu.launch_func @gpu_kernels::@kernel args(%mem : memref<5xf32>)
    return
  }
  gpu.module @gpu_kernels {
    gpu.func @kernel(%arg0 : memref<5xf32, #gpu.address_space<global>>) kernel {
      gpu.return
    }
  }
```

The correct address space for a kernel argument is `1` for global (OpenCL's requirement), but it doesn't make any sense on the host side, for which, `0` is the right one (say, we rely on some runtime mechanism to deliver the data to the device). The two don't match and validation fails on the type checking even though the code is still valid.

The easiest workaround we discussed with @victor-eds is to allow this discrepancy on the validation side. It can be even more specific, and check the target to ensure this is the right case.

### CODE REVIEW

File: mlir/lib/Dialect/GPU/IR/GPUDialect.cpp [hunk 1]
```diff
@@ -401,8 +401,29 @@ LogicalResult GPUDialect::verifyOperationAttribute(Operation *op,
...
+      // Allow address space incompatibility for OpenCL kernels: `gpu.launch`'s
+      // argument memref without address space attribute will match a kernel
+      // function's memref argument with address space `Global`.
+      if (launchOpMemref && kernelMemref) {
+        auto launchAS = llvm::dyn_cast_or_null<gpu::AddressSpaceAttr>(
+            launchOpMemref.getMemorySpace());
+        auto kernelAS = llvm::dyn_cast_or_null<gpu::AddressSpaceAttr>(
+            kernelMemref.getMemorySpace());
+        if (!launchAS && kernelAS &&
+            kernelAS.getValue() == gpu::AddressSpace::Global)
```
Comment (victor-eds) [inline mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:416]: Can't you also pass a pointer to workgroup (aka local, shared) memory here from the host? Can we simply not check memory space at all between launchpoint and kernel?

Comment (kurapov-peter) [reply mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:416]: Yup, that's what I started with. I wasn't sure how specific the check should be, so just put this as an example. We can either go with skipping them altogether or describe the cases precisely.

Comment (kurapov-peter) [reply mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:416]: > Can't you also pass a pointer to workgroup (aka local, shared) memory here from the host?

Not sure, what would it mean? I.e. what's the expected behavior?

Comment (victor-eds) [reply mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:416]: According to the OpenCL spec  https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#setting-kernel-arguments:

```OpenCL
// Provided by CL_VERSION_1_0
cl_int clSetKernelArg(
    cl_kernel kernel,
    cl_uint arg_index,
    size_t arg_size,
    const void* arg_value);
```

> arg_size specifies the size of the argument value. If the argument is a memory object, the arg_size value must be equal to sizeof(cl_mem). *For arguments declared with the local qualifier, the size specified will be the size in bytes of the buffer that must be allocated for the local argument.* If the argument is of type sampler_t, the arg_size value must be equal to sizeof(cl_sampler). If the argument is of type queue_t, the arg_size value must be equal to sizeof(cl_command_queue). For all other arguments, the size will be the size of argument type.


Also, in the OpenCL-SPIR-V client specification https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Env.html#_kernel_arguments:

> For OpTypePointer parameters, supported Storage Classes are:
> - CrossWorkgroup
> - Workgroup
> - UniformConstant


I would also allow SLM pointer arguments here, as that's how, e.g., SYCL is doing it AFAIL

Comment (kurapov-peter) [reply mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:416]: Thanks for the links!

> According to the OpenCL spec https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#setting-kernel-arguments

I mean the SLM pointer passed from host would be `null` (and having just one pointer is meaningless anyway since each WG would have its own); this is also in the spec, same section:
> If the argument is declared with the local qualifier, the arg_value entry must be NULL.

We can allow it, but I don't see any practical sense. If a kernel has a parameter marked as local, then the caller should already know about it. In other words, the memory pointer (let's forget that it is NULL for a moment) that is passed to the kernel does not come from a runtime-managed buffer that can migrate between devices. Hence, no such problem should exist for those.

Current definition of the address space enum:
```
def GPU_AddressSpaceEnum : GPU_I32Enum<
  "AddressSpace", "GPU address space", [
    GPU_AddressSpaceGlobal,
    GPU_AddressSpaceWorkgroup,
    GPU_AddressSpacePrivate
  ]>;
```
I suppose for constants it should still have address space global and retrieve constness in the lowering?

P.s. the only reason I'm suggesting to make the check specific at all is that the check can actually be useful I assume, so it seems natural to try and preserve it for all the other cases.

Comment (victor-eds) [reply mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:416]: I see. In fact, I was thinking on a nonexistent issue. I think we're fine, as, when the kernel is still a `gpu.func`, local memory should be represented as memory attributions in the function, so, instead of:
```
gpu.func @kernel(%slm: memref<f32, #gpu.address_space<workgroup>>, ...)
```
we'll have:
```
gpu.func @kernel(...)
    workgroup(%slm: memref<f32, #gpu.address_space<workgroup>>)
```
And only when converting to LLMV dialect (no signature check anymore), we'll have:
```
llvm.func @kernel(..., %slm: llvm.ptr<3>)
```

TL;DR: We are fine. Thanks!

**Review by victor-eds — APPROVED** [review]
LGTM. I'll add more reviewers

**Review by fabianmcg — CHANGES_REQUESTED** [review]
I'm blocking preemptively before discussing the changes, as I'm not that familiar with `spv` design.

Why don't you pass the correct address space from the beginning? For example:
```mlir
func.func @foo(%mem : memref<5xf32, #gpu.address_space<global>>) {
    gpu.launch_func @gpu_kernels::@kernel args(%mem : memref<5xf32, #gpu.address_space<global>>)
    return
  }
  gpu.module @gpu_kernels {
    gpu.func @kernel(%arg0 : memref<5xf32, #gpu.address_space<global>>) kernel {
      gpu.return
    }
  }
```

Also, you need to change the description and title, this is not `LaunchOp`,  but `LaunchFuncOp`

**Comment by fabianmcg** [comment]
> for memory that is shared between host and device (the runtime controls the necessary transfers) the host memref should have address space zero as a regular memory buffer. It's not a gpu buffer, so the attribute doesn't make any sense.

My issue here is that you're changing the semantics of the op for a very particular case. The semantics of `gpu.launch_func` are similar to those of a `call` operation, where the signature is expected to be an exact match.

To me the correct solution is having a new operation for your runtime.

**Comment by kurapov-peter** [comment]
> To me the correct solution is having a new operation for your runtime.

Do you mean something like `gpu.launch_opencl_func`? It's not like it is specific to opencl runtime though; sycl has unified shared memory (see for example here: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:memory.model) - would have the same problem. The semantics is the same too with the address space exception, so the operation would essentially be a copy of the `gpu.launch_func` with the changed validation.

There was also an idea to leave the memrefs intact (no attributes) and then attach address spaces during the lowering to llvm. The problem is that it is unclear what should happen to a kernel that has two pointers as arguments: one with an attribute and one without.

> My issue here is that you're changing the semantics of the op for a very particular case.

Yeah, I thought, maybe making it very specific to the use case would be tolerable. To keep all the other cases intact.

**Comment by fabianmcg** [comment]
> It's not like it is specific to opencl runtime though; sycl has unified shared memory (see for example here: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:memory.model) -

The issue is not unified memory. For example, CUDA and HIP also have it, but their models take the generic address space as the global one, hence no issues arise. 

> The semantics is the same too with the address space exception, so the operation would essentially be a copy of the `gpu.launch_func` with the changed validation.

The address space is part of the type, making exceptions is not the best solution.

> The problem is that it is unclear what should happen to a kernel that has two pointers as arguments: one with an attribute and one without.

You know that you can attach attributes to individual arguments?

One option is adding: `#gpu.address_space<unified>`.

**Comment by victor-eds** [comment]
After taking a look at the MLIR Vulkan runner (also SPIR-V-based), it faces a similar issue and it's solution is to run a pass to set the expected address spaces both in host and device as part of the conversion pipeline. I haven't dived deep enough into this, but my first perception is this works "by chance" (or good planning in advance!).

The address space they are assuming for default (`memref<f32>`) memory spaces is `StorageBuffer`. So, if they have:
```
gpu.func @kernel(%arg0: memref<f32>) ...
...
gpu.launch @kernel(%arg0: memref<f32>)
```

This would  be changed to:

```
gpu.func @kernel(%arg0: memref<f32, #spirv.StorageClass<StorageBuffer>>) ...
...
gpu.launch @kernel(%arg0: memref<f32, #spirv.StorageClass<StorageBuffer>>)
```

And, when lowered to LLVM dialect (pseudocode):

```
llvm.func @kernel(%arg0: !llvm.ptr) ...
...
"launch_op" @kernel(%arg0: !llvm.ptr)
```

It works fine because the `StorageBuffer` storage class is converted to address space 0 in LLVM. On the host, these programming models (AFAIK) use the default (0) address space both for host-side allocations and device-side (needed for this parameter passing). It happens to be the case the device side uses LLVM address space 0 to represent this particular storage class.

In our case, we'd need to use the `CrossWorkgroup` storage class there, mapping to LLVM address space 1 and, well, we'd be using address space 1 on the host (something these programming models do not do at all). 

IMO being explicit with the memory spaces is better than any other tricks we come up with to avoid this issue. I am not that familiar with other programming models and I don't know how CUDA/HIP pipelines handle this, but, as there are examples of address space mismatches between host and device in LLVM, I would reconsider the approach proposed by @kurapov-peter here.

**Comment by kurapov-peter** [comment]
> The issue is not unified memory. For example, CUDA and HIP also have it, but their models take the generic address space as the global one, hence no issues arise.

Yeah... somebody decided `1` is a great default value, now I'm suffering. One thought that comes to mind is whether this should be a part of some kind of a host-device calling convention (which is runtime-specific). That would make the validation strict by means of adhering to the convention and not rely on the default values.

> After taking a look at the MLIR Vulkan runner (also SPIR-V-based), it faces a similar issue and it's solution is to run a pass to set the expected address spaces both in host and device as part of the conversion pipeline.

Oh, interesting, thanks Victor! Isn't that weird though? I suppose it also relies on the implicit knowledge that on host this would yield correct address space just because the default one is zero?

**Comment by kurapov-peter** [comment]
> > The problem is that it is unclear what should happen to a kernel that has two pointers as arguments: one with an attribute and one without.
> > 
> You know that you can attach attributes to individual arguments?

Sure, I mean it is unclear whether the user's intention actually was to have one of the arguments without the attribute for whatever reason. I suspect this shouldn't be a problem in practice though, since you need `addrpspace(1)` in any case.

> One option is adding: #gpu.address_space<unified>.

That's interesting. Probably not in gpu too? Like, for all the devices on a platform?

**Comment by fabianmcg** [comment]
> but, as there are examples of address space mismatches between host and device in LLVM, I would reconsider the approach proposed by @kurapov-peter here.

Those mismatches actually have a `TODO` attached see: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/GPU/IR/GPUDialect.cpp#L389-L394 
We shouldn't aggravate the situation.


> That's interesting. Probably not in gpu too? Like, for all the devices on a platform?

What I'm saying, is, if we add a 4th GPU address space: `unified`, that could fit your needs for the SPIR-V-like model.

**Comment by kurapov-peter** [comment]
I'm inclined to just patch the signature in the pass. Sounds cleanest so far. Will give it some more thought.

**Comment by victor-eds** [comment]
> Those mismatches actually have a `TODO` attached see: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/GPU/IR/GPUDialect.cpp#L389-L394
> We shouldn't aggravate the situation.

I meant between host and device in LLVM *IR*. Anyway, I think we can just give this a thought and come back if we find further issues as @kurapov-peter says.

**Comment by kurapov-peter** [comment]
Closing this as we have a cleaner solution in #105664.
